Fix DTS playback - ask for transcoding when unsupported

rickysixx opened the pull request:
<!--
Ensure your title is short, descriptive, and in the imperative mood (Fix X, Change Y, instead of Fixed X, Changed Y).
For a good inspiration of what to write in commit messages and PRs please review https://chris.beams.io/posts/git-commit/ and our https://docs.jellyfin.org/general/contributing/issues.html page.
-->

**Changes**
<!-- Describe your changes here in 1-5 sentences. -->
Check if DTS audio codec is supported by the device before adding it to the supported codec list.

Probably a similar check should be done also for other "advanced" audio codecs (e.g. Dolby TrueHD).

**Issues**
<!-- Tag any issues that this PR solves here.
ex. Fixes # -->
Fixes #3896, #3868

nielsvanvelzen reviewed (commented):
I'm not sure how this will behave when using an external audio receiver with DTS support on a TV without support. Would it wrongly think it can't playback DTS?

MichaelRUSF commented:
I implemented these changes and it did not work as expected.  It caused all DTS to transcode on my Gen2 Firetv Max which supports DTS. On my TV which also supports DTS it played the file directly. 

```
Stream #0:1(eng): Audio: dts (DTS), 48000 Hz, 5.1(side), fltp, 1536 kb/s (default)
Stream mapping:
  Stream #0:0 -> #0:0 (copy)
  Stream #0:1 -> #0:1 (dts (dca) -> aac (libfdk_aac))
```
Most of the issues I've encountered have been with high-end resolution DT XLL files. And that playback issue has been solved by the transcode after direct play fails twice error handling.

rickysixx commented:
Thank you very much for your test @MichaelRUSF. My current setup does not support DTS in any form so I am not able to test that scenario.

In my case the change that forces transcoding after 2 playback errors does not fix the issue (the player closes after 3 errors when I try to play DTS audio).

Since listing supported MediaCodec does not always work, in the next days I will try to come up with a different solution.

rickysixx commented:
I have pushed another set of changes to check for DTS support in another way.

This code surely need some improvements/refactoring, but first I need to know if it actually works as expected also for DTS-capable hardware (I cannot test this directly).

I have already tested it in case of no DTS-capable hardware and it correctly triggers a transcode.

MichaelRUSF commented:
Re-tested and DTS support is correctly identified for both devices now, no transcoding. High-resolution DTS audio still errors out on the FireTV, but that's not b/c of this PR. 

This change should help out devices that lack DTS support.

rickysixx commented:
Thank you for testing this Michael.

@nielsvanvelzen Could you please take a look at this PR when you have a moment?

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/util/profile/ProfileHelper.kt:
```diff
@@ -160,11 +164,22 @@ object ProfileHelper {
+			return true
+		}
+
+		val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
```
Use the androidx.core-ktx extension

```
context.getSystemService<AudioManager>()
```

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/util/profile/ProfileHelper.kt:
```diff
@@ -160,11 +164,22 @@ object ProfileHelper {
+		}
+
+		val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
+		val outputDevices = audioManager.getDevices(AudioManager.GET_DEVICES_OUTPUTS)
+
+		return outputDevices.any { it.encodings.contains(AudioFormat.ENCODING_DTS) }
```
I'm not really familiar with the AudioManager but from my understanding this doesn't work properly. You're currently checking if _any_ of the connected devices has a DTS encoder. Not if the **currently in-use** audio device has a **DTS decoder**.

I'm also not sure if we should use `MediaCodecList` via `MediaCodecCapabilitiesTest` instead, since we do all our other codec checking with that utility class.

rickysixx replied:
> I'm also not sure if we should use MediaCodecList via MediaCodecCapabilitiesTest instead, since we do all our other codec checking with that utility class.

This indeed was the first implementation, which however didn't work as expected because the `MediaCodecList` only contains the codecs supported by the device itself, ignoring any extra codec supported by a device connected through eARC for example.

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/util/profile/ExoPlayerProfile.kt:
```diff
@@ -26,8 +30,8 @@ import org.jellyfin.apiclient.model.dlna.TranscodingProfile
 	disableVideoDirectPlay: Boolean,
-	isAC3Enabled: Boolean,
-	downMixAudio: Boolean
+	context: Context,
+	userPreferences: UserPreferences,
```
Profile should not have any dependency on context/preferences

nielsvanvelzen commented:
@rickysixx can you address the review comments instead of constantly force-pushing the same changes over and over again?

rickysixx committed:
Add preference to enable DTS audio bitstream

Currently there is no reliable way to check if DTS audio is supported by
the device or by others it's connected to, therefore DTS playback must be
enabled/disabled manually by the user.

`false` is the safest default value for the preference, because it allows
media to be always played without errors on any device, even though it will
trigger a useless transcode for devices that do support DTS.

This commit reverts 38589ef0c7124857803cbf5656505c1932a3bb04, but now
the preference does have a logic.

rickysixx committed:
refactor: use lower camel case for variable name
